oximeter/collector/src/agent.rs

```diff
@@ -892,9 +892,16 @@ mod tests {
 
         assert_eq!(stats.collections.datum.value(), 0);
         assert!(count != 0);
-        assert_eq!(
-            count,
-            collection_count.load(Ordering::SeqCst),
+
+        // The server may have handled a request that we've not yet recorded on
+        // our collection task side, so we allow the server count to be greater
+        // than our own. But since the collection task is single-threaded, it
+        // cannot ever be more than _one_ greater than our count, since we
+        // should increment that counter before making another request to the
+        // server.
```

Comment:
With apologies for my lack of context: I don't follow the argument about the collection task being single-threaded. Is the "off by one" because there's a window between when we get `stats` from the collector and when we read the `server_count` atomic where the collector could send additional requests? How does the collection task being single threaded mean that window allows at most 1?

Is it possible to cleanly shut down the collector? Would it make sense to shut down the collector, get the final stats from it, then compare that to `server_count`? (I'm wondering if we can deterministically close the window where the collector could do more stuff.)

Reply:
I think the sequence where we're off by one goes like this:

- The inner spawned task makes a request to the server
- The server bumps its counter and then responds
- We ask for the collection task's counts, which is one fewer than the server side
- The inner collection task sends its result to the outer collection task, which bumps _its_ counter. They're now equal.

> Is it possible to cleanly shut down the collector?

This is what I was hoping to do, but it's not possible with the API today to shutdown the collector and _then_ get its stats out. We could make the shutdown method return the current stats, possibly?

Reply:
> > Is it possible to cleanly shut down the collector?
> 
> This is what I was hoping to do, but it's not possible with the API today to shutdown the collector and _then_ get its stats out. We could make the shutdown method return the current stats, possibly?

I think this is at least worth a second look - shutdown returning any relevant state seems perfectly reasonable (and particularly useful for tests like this one). If this is hard or messy I wouldn't spend a _ton_ of time on it, but it feels a lot nicer than having to allow for off-by-one tests.

Reply:
I think we'll have the same problem in this case. It's possible that the shutdown message is received and processed _before_ the results are forwarded from the inner collection task to the outer one. We'll still have the possibility off-by-one between the server and the collection task.

Reply:
Oh, yeah, fair point.